### PR TITLE
Git-Trees API.

### PR DESCRIPTION
**Pull request by Kasi-R**
Started on Git-Trees API. Added methods for getting a tree and getting a tree recursively as per https://developer.github.com/v3/git/trees/

Creating a tree not done.

### CODE REVIEW

File: routers/api/v1/repo/tree.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package repo
+
+import (
+	"code.gitea.io/gitea/modules/context"
```
Comment (lafriks) [inline routers/api/v1/repo/tree.go]: Import order wrong

File: routers/api/v1/repo/tree.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	"fmt"
+	"code.gitea.io/gitea/modules/setting"
+	"strings"
+	"code.gitea.io/git"
+)
+
+type TreeEntry struct {
```
Comment (lafriks) [inline routers/api/v1/repo/tree.go]: Types must be reused from sdk package

File: routers/api/v1/api.go [hunk 3]
```diff
@@ -389,6 +392,9 @@ func RegisterRoutes(m *macaron.Macaron) {
 						m.Post("/tests", context.RepoRef(), repo.TestHook)
 					})
 				}, reqToken(), reqRepoWriter())
+
```
Comment (lafriks) [inline routers/api/v1/api.go]: Why so many empty lines here?

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by Kasi-R** [comment]
Done. I used the types from https://github.com/go-gitea/go-sdk/pull/109 however, i can change it to the types used in https://github.com/go-gitea/go-sdk/pull/94 if required.

**Comment by techknowlogick** [comment]
@Kasi-R would it be possible to use other PR in SDK? I think it would be best to merge in functionality then another PR for enhancement.

**Comment by techknowlogick** [comment]
@lafriks please update review. Can you also review go-gitea/go-sdk#94?

File: routers/api/v1/api.go [hunk 4]
```diff
@@ -368,17 +368,16 @@ func RegisterRoutes(m *macaron.Macaron) {
...
-
 		m.Combo("/repositories/:id", reqToken()).Get(repo.GetByID)
 
 		m.Group("/repos", func() {
+			m.Get("/search", repo.Search)
 			m.Post("/migrate", reqToken(), bind(auth.MigrateRepoForm{}), repo.Migrate)
-
 			m.Group("/:username/:reponame", func() {
 				m.Combo("").Get(repo.Get).Delete(reqToken(), repo.Delete)
+				m.Group("/trees", func() {
```
Comment (lafriks) [inline routers/api/v1/api.go]: we should probably use github compatible route (`/git/trees`)

**Comment by lafriks** [comment]
@techknowlogick done

File: routers/api/v1/repo/tree.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	if ctx.QueryBool("recursive") {
+		Entries, err = GitTree.ListEntriesRecursive()
+	} else {
+		Entries, err = GitTree.ListEntries()
+	}
+	if err != nil {
+		return tree
+	}
+	RepoIDLen := len(RepoID)
+	BlobURL := make([]byte, RepoIDLen + 47)
```
Comment (necaris) [inline routers/api/v1/repo/tree.go]: @Kasi-R could you please add a comment or variable describing where the magic number `47` comes from?

File: routers/api/v1/repo/tree.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	copy(BlobURL[:], RepoID)
+	copy(BlobURL[RepoIDLen:], "/blobs/")
+	TreeURL := make([]byte, RepoIDLen + 47)
```
Comment (necaris) [inline routers/api/v1/repo/tree.go]: See above -- same here please.

File: routers/api/v1/repo/tree.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	copy(TreeURL[:], RepoID)
+	copy(TreeURL[RepoIDLen:], "/trees/")
+	CopyPos := len(TreeURL) - 40
```
Comment (necaris) [inline routers/api/v1/repo/tree.go:62]: Similar request here -- where is the `40` coming from?

**Review by necaris — APPROVED** [review]
Just a couple of small requests for better explanation of where some constants are coming from -- otherwise this looks good, very straightforward -- certainly simpler than my attempt #1978 !

**Pull request closed without merging**